contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol

```diff
@@ -358,16 +348,12 @@ abstract contract FunctionsBilling is HasRouter, IFunctionsBilling {
   function _fulfillAndBill(
     bytes32 requestId,
     bytes memory response,
-    bytes memory err
-  )
-    internal
-    returns (
-      /* bytes calldata metadata, */
-      FulfillResult
-    )
-  {
-    Commitment memory commitment = s_requestCommitments[requestId];
-    if (commitment.don == address(0)) {
+    bytes memory err,
+    bytes memory onchainMetadata,
+    bytes memory /* offchainMetadata TODO: use in getDonFee() for dynamic billing */
+  ) internal returns (FulfillResult) {
+    IFunctionsRequest.Commitment memory commitment = _retrieveCommitmentData(onchainMetadata);
+    if (s_requestCommitments[requestId] != keccak256(abi.encode(commitment))) {
       return FulfillResult.INVALID_REQUEST_ID;
     }
     delete s_requestCommitments[requestId];
```

Comment:
I am still not convinced we should be deleting the request commitment until we confirm we have actually called the user's callback with the correct amount of gas. However, since that is outside of the scope of this PR, we can have the discussion elsewhere.

Reply:
Not going to pile that into this PR, that is the concern of https://github.com/smartcontractkit/chainlink/pull/9984